eqcorrscan/scripts/correlation_speeds.py

```diff
@@ -0,0 +1,176 @@
+from multiprocessing import cpu_count
+
+from eqcorrscan.utils.correlate import (
+    XCOR_FUNCS, XCORR_STREAM_METHODS, get_stream_xcorr)
+from eqcorrscan.utils import correlate
+
+
+def memory_limit():
+    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
+    resource.setrlimit(resource.RLIMIT_AS, (get_memory() * 1024, hard))
+
+
+def get_memory():
+    with open('/proc/meminfo', 'r') as mem:
+        free_memory = 0
+        for i in mem:
+            sline = i.split()
+            if str(sline[0]) in ('MemFree:', 'Buffers:', 'Cached:'):
+                free_memory += int(sline[1])
+    return free_memory
+
+
+def generate_dataset(n_templates, n_stations, n_channels, data_len,
+                     template_len, sampling_rate):
+    """
+
+    :param n_templates:
+    :param n_stations:
+    :param n_channels:
+    :param data_len:
+    :param template_len:
+    :param sampling_rate:
+    :return:
+    """
+    print("=" * 80)
+    print("Generating dataset".center(80))
+    print("=" * 80)
+    dataset = {'data': Stream(), 'templates': []}
+    for i in range(n_stations * n_channels):
+        dataset['data'] += Trace(np.random.randn(
+            int(data_len * sampling_rate)))
+    for t in range(n_templates):
+        template = Stream()
+        for i in range(n_stations * n_channels):
+            template += Trace(np.random.randn(
+                int(template_len * sampling_rate)))
+        dataset['templates'].append(template)
+    return dataset
+
+
+def run_correlation(func, threads, dataset, loops=3):
+    """
+    Run the given correlation function and profile time and memory usage.
+
+    :param func:
+    :param threads:
+    :param dataset:
+    :param loops:
+    :return:
+    """
+    max_mem = []
+    timings = []
+    for loop in range(loops):
+        tic = time.time()
+        max_mem = memory_usage((func, (dataset['templates'], dataset['data']),
+                                {'cores': threads}), max_usage=True,
+                               include_children=True, multiprocess=True)
+        toc = time.time()
+        timings.append(toc - tic)
+    return np.mean(max_mem[0]), np.mean(timings)
+
+
+def plot_profiles(times, memory_use):
+    """
+
+    :param times:
+    :param memory_use:
+    :return:
+    """
+    plt.xkcd()
+    markers = itertools.cycle((',', '+', '.', 'o', '*'))
+    for key in times.keys():
+        plt.scatter(memory_use[key], times[key], s=10**2, marker=next(markers),
+                    label=key)
+    plt.xlabel('Memory use (MB)')
+    plt.ylabel('Time (s)')
+    plt.legend()
+    plt.show()
+
+
+def run_profiling(n_templates, n_stations, n_channels, data_len,
+                  template_len, sampling_rate, loops=3):
+    """
+
+    :param n_templates:
+    :param n_stations:
+    :param n_channels:
+    :param data_len:
+    :param template_len:
+    :param sampling_rate:
+    :return:
+    """
+    print("Testing correlation functions in %s" % correlate.__file__)
```

Comment:
Actually you will be testing all registered correlation functions, not just the ones in correlate (although this will probably be true most of the time). Maybe just print the keys of XCOR_FUNCS?

Reply:
This was more in there for my debugging, to make sure I was testing the installed functions rather than those on my local path.